File: javascript/parenth_balance.js
function verify_parenth_balancing(expr) {
    let stack = [];
    for (let i = 0; i < expr.length; i++) {
        if (expr[i] == '(') {
            stack.push('(');
        }
        else if (expr[i] == ')') {
            if (stack.length == 0) {
                 return false;
            }
            else {
                stack.pop();
            }
        }
    }

    return stack.length == 0;
}

expr = '(e^sin(x)) * 2 (';

if (verify_parenth_balancing(expr)) {
    console.log(expr, "is balanced");
}
else {
    console.log(expr, "is unbalanced");
}

File: leetcode/parenth_balance.py
#https://leetcode.com/problems/valid-parentheses/description/

import sys

def verify_parentheses(expr):
  stack = []
  for c in expr:
    if len(stack) == 0 and (c == ')' or c == ']' or c == '}'):
      return False
    if c == '(' or c == '[' or c == '{':
      stack.append(c)
    elif c == ')':
      if stack[len(stack) - 1] == '(':
        stack.pop()
      else:
        return False
    elif c == ']':
      if stack[len(stack) - 1] == '[':
        stack.pop()
      else:
        return False
    elif c == '}':
      if stack[len(stack) - 1] == '{':
        stack.pop()
      else:
        return False

  return len(stack) == 0

print(verify_parentheses(sys.argv[1]))
